File: ios/CISOAlert/Models/Briefing.swift
import Foundation

enum Severity: String, Codable, CaseIterable {
    case critical, high, medium, low, unknown
    var rank: Int { switch self { case .critical: 4; case .high: 3; case .medium: 2; case .low: 1; case .unknown: 0 } }
}

struct BriefingItem: Codable, Identifiable, Hashable {
    let id: String
    let title: String
    let severity: Severity
    let cvss: Double?
    let kev: Bool
    let techKeys: [String]
    let summary: String
    let impact: String
    let action: String
    let published: String
    let sourceURL: String?
    let category: String?
    let sourceName: String?
}

struct BriefingGroup: Codable, Identifiable, Hashable {
    let title: String
    let product: String
    let count: Int
    let maxSeverity: Severity
    let items: [BriefingItem]
    var id: String { product }
}

enum BriefingEntry: Codable, Hashable, Identifiable {
    case item(BriefingItem)
    case group(BriefingGroup)

    var id: String {
        switch self {
        case .item(let i): "item-\(i.id)"
        case .group(let g): "group-\(g.product)"
        }
    }

    private enum CodingKeys: String, CodingKey { case kind, item, title, product, count, maxSeverity, items }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        switch try c.decode(String.self, forKey: .kind) {
        case "item":
            self = .item(try c.decode(BriefingItem.self, forKey: .item))
        case "group":
            self = .group(BriefingGroup(
                title: try c.decode(String.self, forKey: .title),
                product: try c.decode(String.self, forKey: .product),
                count: try c.decode(Int.self, forKey: .count),
                maxSeverity: try c.decode(Severity.self, forKey: .maxSeverity),
                items: try c.decode([BriefingItem].self, forKey: .items)))
        case let other:
            throw DecodingError.dataCorruptedError(forKey: .kind, in: c, debugDescription: "unknown entry kind \(other)")
        }
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .item(let i):
            try c.encode("item", forKey: .kind)
            try c.encode(i, forKey: .item)
        case .group(let g):
            try c.encode("group", forKey: .kind)
            try c.encode(g.title, forKey: .title)
            try c.encode(g.product, forKey: .product)
            try c.encode(g.count, forKey: .count)
            try c.encode(g.maxSeverity, forKey: .maxSeverity)
            try c.encode(g.items, forKey: .items)
        }
    }
}

struct Briefing: Codable, Hashable {
    let date: String
    let headline: String
    let entries: [BriefingEntry]
}

struct BriefingDate: Codable, Identifiable, Hashable {
    let date: String
    let count: Int
    let maxSeverity: Severity
    var id: String { date }
}

enum BriefingCategory: String, CaseIterable {
    case vuln, threat, news
    var title: String { switch self { case .vuln: "Vulns"; case .threat: "Threats"; case .news: "News" } }
    var systemImage: String { switch self { case .vuln: "exclamationmark.shield"; case .threat: "bolt.shield"; case .news: "newspaper" } }
}

File: worker/src/briefing.ts
import type { Item, Briefing, BriefingEntry, Category } from "./types";
import { displayProduct } from "./intel/product";

const sevRank: Record<Item["severity"], number> = { critical: 4, high: 3, medium: 2, low: 1, unknown: 0 };
const GROUP_MIN = 3;

const CONFIG: Record<Category, { group: boolean; global: boolean; limit: number | null }> = {
  vuln:   { group: true,  global: false, limit: null },
  threat: { group: false, global: true,  limit: 15 },
  news:   { group: false, global: true,  limit: 5 }
};

const vulnRank = (a: Item, b: Item) =>
  (Number(b.kev) - Number(a.kev)) || (sevRank[b.severity] - sevRank[a.severity]) || b.published.localeCompare(a.published);
const articleRank = (a: Item, b: Item) =>
  ((b.score ?? 0) - (a.score ?? 0)) || b.published.localeCompare(a.published);

export interface ComposeOpts { date: string; category: Category; stack: string[]; fallback?: Item[]; }

export function composeBriefing(items: Item[], opts: ComposeOpts): Briefing {
  const { date, category, stack, fallback = [] } = opts;
  const cfg = CONFIG[category];
  const stackSet = new Set(stack);
  const rank = category === "vuln" ? vulnRank : articleRank;

  let matched = cfg.global ? items.slice() : items.filter(i => i.techKeys.some(k => stackSet.has(k)));

  // For vulns: merge global KEV/critical fallback so actively-exploited items surface
  // even when they don't match the user's stack.
  if (category === "vuln" && fallback.length) {
    const have = new Set(matched.map(i => i.id));
    matched = matched.concat(fallback.filter(i => !have.has(i.id)));
  }

  matched.sort(rank);
  if (cfg.limit != null) matched = matched.slice(0, cfg.limit);

  const entries = cfg.group ? groupedEntries(matched) : matched.map(i => {
    const { source, ...item } = i; return { kind: "item", item } as BriefingEntry;
  });

  return { date, headline: headlineFor(category, matched.length), entries };
}

function groupedEntries(matched: Item[]): BriefingEntry[] {
  const productCount = new Map<string, number>();
  for (const it of matched) { const p = (it.product ?? "").toLowerCase(); if (p) productCount.set(p, (productCount.get(p) ?? 0) + 1); }
  const collapse = new Set([...productCount].filter(([, c]) => c >= GROUP_MIN).map(([p]) => p));
  const emitted = new Set<string>();
  const entries: BriefingEntry[] = [];
  for (const it of matched) {
    const p = (it.product ?? "").toLowerCase();
    if (p && collapse.has(p)) {
      if (emitted.has(p)) continue;
      emitted.add(p);
      const members = matched.filter(m => (m.product ?? "").toLowerCase() === p).map(({ source, ...rest }) => rest);
      entries.push({ kind: "group", title: `Multiple ${displayProduct(it.product!)} Security Issues`, product: p, count: members.length, maxSeverity: members[0].severity, items: members });
    } else {
      const { source, ...item } = it;
      entries.push({ kind: "item", item });
    }
  }
  return entries;
}

function headlineFor(category: Category, n: number): string {
  if (n === 0) return category === "vuln" ? "No issues match your stack for this briefing"
    : category === "threat" ? "No major threats match your stack today"
    : "No AI & cyber news for this date";
  if (category === "vuln") return `${n} issue${n === 1 ? "" : "s"} match your stack`;
  if (category === "threat") return `${n} threat${n === 1 ? "" : "s"} to know today`;
  return `Today's top ${n} AI & cyber stor${n === 1 ? "y" : "ies"}`;
}
